src/rust/src/zip339_ffi.rs

```diff
@@ -0,0 +1,110 @@
+
+use zcash_primitives::zip339;
+
+#[repr(u32)]
+#[derive(Copy, Clone)]
+pub enum Language {
+    English = 0,
+    SimplifiedChinese = 1,
+    TraditionalChinese = 2,
+    Czech = 3,
+    French = 4,
+    Italian = 5,
+    Japanese = 6,
+    Korean = 7,
+    Portuguese = 8,
+    Spanish = 9,
+}
+
+impl From<Language> for zip339::Language {
+    fn from(language: Language) -> Self {
+        match language {
+            Language::English => zip339::Language::English,
+            Language::SimplifiedChinese => zip339::Language::SimplifiedChinese,
+            Language::TraditionalChinese => zip339::Language::TraditionalChinese,
+            Language::Czech => zip339::Language::Czech,
+            Language::French => zip339::Language::French,
+            Language::Italian => zip339::Language::Italian,
+            Language::Japanese => zip339::Language::Japanese,
+            Language::Korean => zip339::Language::Korean,
+            Language::Portuguese => zip339::Language::Portuguese,
+            Language::Spanish => zip339::Language::Spanish,
+        }
+    }
+}
+
+/// Creates a phrase with the given entropy, in the given `Language`. The phrase is represented as
+/// a pointer to a null-terminated C string that must not be written to, and must be freed by
+/// `zip339_free_phrase`.
+#[no_mangle]
+pub extern "C" fn zip339_entropy_to_phrase(
+    language: Language,
+    entropy: *const u8,
+    entropy_len: size_t,
+) -> *const c_uchar {
+    assert!(!entropy.is_null());
+    let entropy = unsafe { slice::from_raw_parts(entropy, entropy_len) }.to_vec();
+
+    if let Ok(mnemonic) = zip339::Mnemonic::from_entropy_in(language.into(), entropy) {
+        let phrase = mnemonic.phrase();
+        let phrase_len = phrase.len();
+        let phrase_cstr =
+            unsafe { libc::calloc((phrase_len + 1) as size_t, 1 as size_t) as *mut c_uchar };
+        assert!(!phrase_cstr.is_null());
+        unsafe { ptr::copy(phrase.as_ptr(), phrase_cstr, phrase_len) };
+        phrase_cstr
+    } else {
+        ptr::null()
+    }
+}
+
+/// Frees a phrase returned by `zip339_entropy_to_phrase`.
+#[no_mangle]
+pub extern "C" fn zip339_free_phrase(phrase: *const c_uchar) {
+    if !phrase.is_null() {
+        unsafe {
+            libc::free(phrase as *mut c_void);
+        }
+    }
+}
+
+/// Returns `true` if the given string is a valid mnemonic phrase in the given `Language`.
+#[no_mangle]
+pub extern "C" fn zip339_validate_phrase(language: Language, phrase: *const c_uchar) -> bool {
+    assert!(!phrase.is_null());
+    let phrase = unsafe { CStr::from_ptr(phrase as *const c_char) }
+        .to_str()
+        .unwrap();
+    zip339::Mnemonic::validate_in(language.into(), phrase).is_ok()
+}
+
+/// Copies the seed for the given phrase into the 64-byte buffer `buf`.
+/// Returns true if successful, false on error. In case of error, `buf` is zeroed.
+#[no_mangle]
+pub extern "C" fn zip339_phrase_to_seed(
+    language: Language,
```

Comment:
It's unfortunate that the language is needed here. It seems to be because `bip0039-rs` always validates the phrase, which you can't do without the language.